components/style/restyle_hints.rs

```diff
@@ -383,38 +417,73 @@ impl DependencySet {
+        let state_changes = snapshot.state()
+                                    .map_or_else(ElementState::empty, |old_state| current_state ^ old_state);
         let attrs_changed = snapshot.has_attrs();
+
+        if state_changes.is_empty() && !attrs_changed {
+            return RestyleHint::empty();
+        }
+
         let mut hint = RestyleHint::empty();
-        for dep in &self.deps {
-            if state_changes.intersects(dep.sensitivities.states) || (attrs_changed && dep.sensitivities.attrs) {
-                let old_el: ElementWrapper<E> = ElementWrapper::new_with_snapshot(el.clone(), snapshot);
+        let snapshot = ElementWrapper::new_with_snapshot(el.clone(), snapshot);
+
+        Self::compute_partial_hint(&self.common_deps, el, &snapshot,
+                                   &state_changes, attrs_changed, &mut hint);
+
+        if !state_changes.is_empty() && !hint.is_all() {
+            Self::compute_partial_hint(&self.state_deps, el, &snapshot,
+                                       &state_changes, attrs_changed, &mut hint);
+        }
+
+        if attrs_changed && !hint.is_all() {
+            Self::compute_partial_hint(&self.attr_deps, el, &snapshot,
+                                       &state_changes, attrs_changed, &mut hint);
+        }
+
+        hint
+    }
+
+    fn compute_partial_hint<E>(deps: &[Dependency],
+                               element: &E,
+                               snapshot: &ElementWrapper<E>,
+                               state_changes: &ElementState,
+                               attrs_changed: bool,
+                               hint: &mut RestyleHint)
+        where E: ElementExt
+    {
+        for dep in deps {
+            debug_assert!(state_changes.intersects(dep.sensitivities.states) ||
+                          attrs_changed && dep.sensitivities.attrs,
+                          "Testing a completely ineffective dependency?");
+            if !hint.intersects(dep.hint) {
                 let matched_then =
-                    matches_complex_selector(&*dep.selector, &old_el, None, &mut StyleRelations::empty());
+                    matches_complex_selector(&dep.selector, snapshot, None,
+                                             &mut StyleRelations::empty());
                 let matches_now =
-                    matches_complex_selector(&*dep.selector, el, None, &mut StyleRelations::empty());
+                    matches_complex_selector(&dep.selector, element, None,
+                                             &mut StyleRelations::empty());
                 if matched_then != matches_now {
-                    hint.insert(combinator_to_restyle_hint(dep.combinator));
-                    if hint.is_all() {
-                        break
-                    }
+                    hint.insert(dep.hint);
                 }
             }
+            if hint.is_all() {
```

Comment:
Nit: Seems like this should go inside the branch where we actually modify hint, right?